Documentation/grpc-metadata.md

```diff
@@ -0,0 +1,223 @@
+# Metadata
+
+gRPC supports sending metadata between client and server.
+This doc shows how to send and receive metadata in gRPC-go.
+
+## Background
+
+Four kinds of service method:
+
+- [Unary RPC](http://www.grpc.io/docs/guides/concepts.html#unary-rpc)
+- [Server streaming RPC](http://www.grpc.io/docs/guides/concepts.html#server-streaming-rpc)
+- [Client streaming RPC](http://www.grpc.io/docs/guides/concepts.html#client-streaming-rpc)
+- [Bidirectional streaming RPC](http://www.grpc.io/docs/guides/concepts.html#bidirectional-streaming-rpc)
+
+And concept of [metadata](http://www.grpc.io/docs/guides/concepts.html#metadata).
+
+## Contructing metadata
+
+A metadata can be created using package [metadata](https://godoc.org/google.golang.org/grpc/metadata).
+The type MD is actually a map from string to a list of strings:
+
+```go
+type MD map[string][]string
+```
+
+The content of metadata can be accessed as reading normal maps.
+Note that the value type of this map is `[]string`.
+
+### Creating a new metadata
+
+A metadata can be created from a `map[string]string` using function `New`:
+
+```go
+md := metadata.New(map[string]string{"key1": "val1", "key2": "val2"})
+```
+
+Another way is to use `Pairs`.
+Values with the same key will be merged into a list:
+
+```go
+md := metadata.Pairs("key1", "val1", "key2", "val2")
+```
+
+__Note:__ all keys in metadata will be lowercased.
```

Comment:
should be rather than will be? (or gRPC will automatically convert uppercase to lowercase?)